hail/src/main/scala/is/hail/backend/LowerTableIR.scala

```diff
@@ -1,71 +1,61 @@
+class LowererUnsupportedOperation(msg: String = null) extends Exception(msg)
 
-case class SparkShuffle(child: SparkStage)
+case class ShuffledStage(child: TableStage)
 
-case class SparkBinding(name: String, value: IR)
+case class Binding(name: String, value: IR)
 
-case class SparkStage(
-  globals: SparkBinding,
-  otherBroadcastVals: List[SparkBinding],
+case class TableStage(
+  broadcastVals: IR,
+  globalsField: String,
   rvdType: RVDType,
   partitioner: RVDPartitioner,
-  rdds: Map[String, RDD[RegionValue]],
   contextType: Type,
   contexts: IR,
   body: IR) {
 
-  val broadcastVals: List[SparkBinding] = otherBroadcastVals :+ globals
+  def toIR(bodyTransform: IR => IR): CollectDistributedArray =
+    CollectDistributedArray(contexts, broadcastVals, "context", "global", bodyTransform(body))
 
-  def toIR(bodyTransform: IR => IR): CollectDistributedArray = {
-    val globalVals = MakeStruct(broadcastVals.map { case SparkBinding(n, v) => n -> v })
-    val substEnv = Env[IR](broadcastVals.map(b => b.name -> GetField(Ref("global", globalVals.typ), b.name)): _*)
-    val newBody = Subst(bodyTransform(body), BindingEnv(substEnv))
-    CollectDistributedArray(contexts, globalVals, "context", "global", newBody)
-  }
+  def broadcastRef: IR = Ref("global", broadcastVals.typ)
+  def contextRef: IR = Ref("context", contextType)
+  def globals: IR = GetField(broadcastRef, globalsField)
 }
 
 object LowerTableIR {
 
-  def apply(ir0: IR, timer: ExecutionTimer, optimize: Boolean = true): IR = {
+  def apply(ir0: IR, timer: Option[ExecutionTimer], optimize: Boolean = true): IR = {
+
+    def opt(context: String, ir: IR): IR =
+      Optimize(ir, noisy = true, canGenerateLiterals = true,
+        Some(timer.map(t => s"${ t.context }: $context")
+          .getOrElse(context)))
+
+    def time(context: String, ir: (String) => IR): IR =
+      timer.map(t => t.time(ir(context), context))
+        .getOrElse(ir(context))
+
     var ir = ir0
 
     ir = ir.unwrap
-    if (optimize) {
-      val context = "first pass"
-      ir = timer.time(
-        Optimize(ir, noisy = true, canGenerateLiterals = true, Some(s"${ timer.context }: $context")),
-        context)
-    }
-
-    ir = timer.time(LiftNonCompilable(ir).asInstanceOf[IR], "lifting non-compilable")
-    ir = timer.time(LowerMatrixIR(ir), "lowering MatrixIR")
-
-    if (optimize) {
-      val context = "after MatrixIR lowering"
-      ir = timer.time(
-        Optimize(ir, noisy = true, canGenerateLiterals = true, Some(s"${ timer.context }: $context")),
-        context)
-    }
-
-    ir = LowerTableIR.lower(ir)
-
-    if (optimize) {
-      val context = "after TableIR lowering"
-      ir = timer.time(
-        Optimize(ir, noisy = true, canGenerateLiterals = true, Some(s"${ timer.context }: $context")),
-        context)
-    }
+    if (optimize) { ir = time( "first pass", opt(_, ir)) }
+
+    ir = time("lifting non-compilable", _ => LiftNonCompilable(ir).asInstanceOf[IR])
```

Comment:
I've also removed this step from the lowering since it should just error once it hits a node that isn't compilable or lowerable.